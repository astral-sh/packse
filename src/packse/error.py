import textwrap
from pathlib import Path


class PackseError(Exception):
    """Base type for all Packse exceptions"""


class UserError(PackseError):
    """It's the user's fault :)"""


class RequiresExtra(UserError):
    def __init__(self, feature: str, extra: str) -> None:
        message = f"Using {feature} requires installation with extra {extra!r}"
        super().__init__(message)


class DestinationAlreadyExists(UserError):
    def __init__(self, destination: Path) -> None:
        message = f"Destination directory '{destination}' already exists"
        super().__init__(message)


class InvalidScenario(UserError):
    def __init__(self, path: Path, reason: str) -> None:
        message = f"File at '{path}' is not a valid scenario: {reason}"
        super().__init__(message)


class InvalidPackageVersion(UserError):
    """Version is not PEP compliant"""

    def __init__(self, version: str) -> None:
        message = f"Version {version!r} is not valid."
        super().__init__(message)


class FileNotFound(UserError):
    def __init__(self, path: Path) -> None:
        message = f"File '{path}' not found"
        super().__init__(message)


class BuildError(PackseError):
    def __init__(self, message, output) -> None:
        message = message + ":\n" + textwrap.indent(output, " " * 4)
        super().__init__(message)


class PublishError(PackseError):
    pass


class PublishToolError(PublishError):
    def __init__(self, message, output) -> None:
        message = message + ":\n" + textwrap.indent(output, " " * 4)
        super().__init__(message)


class PublishNoCredentials(PublishError):
    def __init__(self) -> None:
        message = (
            "No credentials found for publish!\n"
            "Provide an API token via `PACKSE_PUBLISH_PASSWORD` or disable authentication with `--anonymous`"
        )
        super().__init__(message)


class PublishAlreadyExists(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish for '{package}' already exists"
        super().__init__(message)


class PublishRateLimit(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish of '{package}' failed due to rate limits"
        super().__init__(message)


class PublishConnectionError(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish of '{package}' failed due to a connection error"
        super().__init__(message)


class InvalidPublishTarget(UserError):
    def __init__(self, path: Path, reason: str) -> None:
        message = f"Publish target at '{path}' is not a valid: {reason}"
        super().__init__(message)


class ServeError(PackseError):
    def __init__(self, message: str) -> None:
        super().__init__(message)


class ServeThreadError(ServeError):
    def __init__(self, message: str) -> None:
        super().__init__(message)


class ServeCommandError(ServeError):
    def __init__(self, message: str, stderr: str) -> None:
        if stderr.strip():
            message += f":\n\n{stderr}"
        super().__init__(message)


class ServeAddressInUse(ServeError):
    def __init__(self, url: str) -> None:
        message = f"Address '{url}' already in use"
        super().__init__(message)


class ServeAlreadyRunning(ServeError):
    def __init__(self) -> None:
        message = "Server is already running"
        super().__init__(message)
