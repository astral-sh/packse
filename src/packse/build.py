"""
Build packages for the given scenarios.
"""
import logging
import shutil
import subprocess
import textwrap
import time
from concurrent.futures import ThreadPoolExecutor
from concurrent.futures import wait as wait_for_futures
from pathlib import Path
from typing import Generator

import packaging.version

from packse.error import (
    BuildError,
    DestinationAlreadyExists,
    FileNotFound,
    InvalidPackageVersion,
    InvalidScenario,
)
from packse.scenario import (
    Package,
    PackageMetadata,
    Scenario,
    load_scenarios,
    scenario_hash,
)
from packse.template import TemplateConfig, create_from_template, load_template_config

logger = logging.getLogger(__name__)


def build(
    targets: list[Path],
    rm_destination: bool,
    short_names: bool,
    no_hash: bool,
    skip_root: bool,
    dist_dir: Path | None = None,
    build_dir: Path | None = None,
):
    start_time = time.time()

    # Validate and collect all targets first
    scenarios = []
    dist_dir = dist_dir or (Path.cwd() / "dist")
    build_dir = build_dir or (Path.cwd() / "build")

    for target in targets:
        if not target.exists():
            raise FileNotFound(target)

        if target.is_dir():
            for target in target.glob("*.json"):
                try:
                    logger.debug("Loading %s", target)
                    scenarios.extend(load_scenarios(target))
                except Exception as exc:
                    invalid = InvalidScenario(target, reason=str(exc))
                    print(f"Skipping file: {invalid}")
        else:
            try:
                logger.debug("Loading %s", target)
                scenarios.extend(load_scenarios(target))
            except Exception as exc:
                raise InvalidScenario(target, reason=str(exc)) from exc

    # Then build each one
    with ThreadPoolExecutor(thread_name_prefix="packse-scenario-") as executor:
        futures = [
            executor.submit(
                build_scenario,
                scenario,
                rm_destination,
                short_names,
                no_hash,
                skip_root,
                build_dir,
                dist_dir,
            )
            for scenario in scenarios
        ]

        wait_for_futures(futures)

    results = [future.result() for future in futures]
    for result in sorted(results):
        print(result)

    logger.info(
        "Built %s scenarios in %.2fs",
        len(results),
        time.time() - start_time,
    )


def build_scenario(
    scenario: Scenario,
    rm_destination: bool,
    short_names: bool,
    no_hash: bool,
    skip_root: bool,
    build_dir: Path,
    dist_dir: Path,
) -> str:
    """
    Build the scenario defined at the given path.

    Returns the scenario's entrypoint package name.
    """

    hash = scenario_hash(scenario)
    name = scenario.name
    if not no_hash:
        name = f"{name}-{hash}"

    build_destination = build_dir / name
    dist_destination = dist_dir / name
    start_time = time.time()

    logger.info(
        "Building '%s' in directory '%s'",
        name,
        build_destination,
    )

    if build_destination.exists():
        if rm_destination:
            shutil.rmtree(build_destination)
        else:
            raise DestinationAlreadyExists(build_destination)

    build_destination.mkdir(parents=True)

    if dist_destination.exists():
        if rm_destination:
            shutil.rmtree(dist_destination)
        else:
            raise DestinationAlreadyExists(dist_destination)

    dist_destination.mkdir(parents=True)

    with ThreadPoolExecutor(thread_name_prefix="packse-build-") as executor:
        futures = [
            executor.submit(
                build_scenario_package,
                scenario=scenario,
                scenario_version=hash,
                name=name,
                package=package,
                build_destination=build_destination,
                dist_destination=dist_destination,
                short_names=short_names,
                no_hash=no_hash,
            )
            for name, package in scenario.packages.items()
        ]

        if not skip_root:
            futures.append(
                executor.submit(
                    build_scenario_package,
                    scenario=scenario,
                    scenario_version=hash,
                    name=scenario.name,
                    package=make_root_package(scenario),
                    build_destination=build_destination,
                    dist_destination=dist_destination,
                    short_names=short_names,
                    no_hash=no_hash,
                )
            )

        wait_for_futures(futures)
        for future in futures:
            # TODO(zanieb): Display _all_ of the errors to the user
            future.result()

    logger.info(
        "Built scenario '%s' in %.2fs",
        name,
        time.time() - start_time,
    )
    return name


def make_root_package(scenario: Scenario) -> Package:
    """
    Generate a full package from the root package of the scenario.
    """
    return Package(
        versions={
            "0.0.0": PackageMetadata(
                requires=scenario.root.requires,
                requires_python=scenario.root.requires_python,
                # Do not build wheels for the root package
                wheel=False,
                # The scenario's description is used for the entrypoint package
                description=scenario.description,
            )
        }
    )


def build_scenario_package(
    scenario: Scenario,
    scenario_version: str,
    name: str,
    package: Package,
    build_destination: Path,
    dist_destination: Path,
    short_names: bool,
    no_hash: bool,
):
    package_name = name
    if not no_hash:
        package_name = f"{package_name}-{scenario_version}"
    if not short_names and name != scenario.name:
        package_name = f"{scenario.name}-{package_name}"

    # Generate a Python module name
    module_name = package_name.replace("-", "_")

    template_config = load_template_config(scenario.template)

    for version, package_version in package.versions.items():
        start_time = time.time()

        logger.debug("Generating project for '%s-%s'", package_name, version)
        package_destination = create_from_template(
            build_destination,
            template_name=scenario.template,
            variables={
                "scenario-name": scenario.name,
                "package-name": package_name,
                "module-name": module_name,
                "version": version,
                "dependencies": [
                    requirement.with_unique_name(
                        scenario, scenario_version, short_names, no_hash
                    )
                    for requirement in package_version.requires
                ],
                "optional-dependencies": [
                    {
                        "name": extra,
                        "dependencies": [
                            requirement.with_unique_name(
                                scenario, scenario_version, short_names, no_hash
                            )
                            for requirement in depends
                        ],
                    }
                    for extra, depends in package_version.extras.items()
                ],
                "requires-python": package_version.requires_python,
                "description": package_version.description,
            },
        )

        logger.info(
            "Generated project for '%s-%s' in %.2fms",
            package_name,
            version,
            (time.time() - start_time) * 1000.0,
        )

        for dist in build_package_distributions(
            template_config, package_version, package_destination
        ):
            shared_path = dist_destination / dist.name
            logger.debug("Linked distribution to %s", shared_path)
            shared_path.hardlink_to(dist)

        if package_version.yanked:
            with dist_destination.with_suffix(".yanked").open("a+") as yanked:
                yanked.write(f"{package_name}-{version}")
                yanked.write("\n")
            logger.info("Marked %s-%s as yanked", package_name, version)


def build_package_distributions(
    template_config: TemplateConfig, package_version: PackageMetadata, target: Path
) -> Generator[Path, None, None]:
    """
    Build package distributions, yield each built distribution path, then delete the distribution folder.
    """
    command = template_config.build_base

    if package_version.sdist:
        command.extend(template_config.build_sdist)
    if package_version.wheel:
        command.extend(template_config.build_wheel)

    start_time = time.time()

    try:
        output = subprocess.check_output(command, cwd=target, stderr=subprocess.STDOUT)

    except subprocess.CalledProcessError as exc:
        raise BuildError(
            f"Building {target.name} with hatch failed",
            exc.output.decode(),
        )
    else:
        logs = (
            (":\n\n" + textwrap.indent(output.decode(), " " * 4))
            if logger.getEffectiveLevel() <= logging.DEBUG
            else ""
        )
        logger.info(
            "Built package '%s' in %.2fs%s",
            target.name,
            time.time() - start_time,
            logs,
        )

    # Create wheels with other tags if requested
    if package_version.wheel and package_version.wheel_tags:
        default_tag_wheel = None
        for dist in (target / "dist").iterdir():
            if dist.name.endswith("py3-none-any.whl"):
                default_tag_wheel = dist
                break
        if not default_tag_wheel:
            raise BuildError("No wheel found with tag `py3-none-any`.")

        # Since we always build a universal wheel, we just create copies of it for other platforms
        # which means we're lying about the compatibility of the wheel but it will always be installable
        # on the given platform.
        for tag in package_version.wheel_tags:
            if tag == "py3-none-any":
                continue

            new_tag_wheel = (
                target / "dist" / default_tag_wheel.name.replace("py3-none-any", tag)
            )
            new_tag_wheel.hardlink_to(default_tag_wheel)
            logger.debug("Created wheel %s", new_tag_wheel.name)

        # Delete the default wheel if not requested
        if "py3-none-any" not in package_version.wheel_tags:
            default_tag_wheel.unlink()

    yield from sorted((target / "dist").iterdir())
    shutil.rmtree(target / "dist")


def build_package(
    name: str,
    version: str,
    no_wheel: bool,
    no_sdist: bool,
    requires_python: str | None,
    wheel_tags: list[str],
    rm_destination: bool,
):
    """
    Build a package without a scenario
    """

    try:
        packaging.version.Version(version)
    except Exception:
        raise InvalidPackageVersion(version) from None

    work_dir = Path.cwd()
    build_destination = work_dir / "build" / name
    dist_destination = work_dir / "dist" / name
    start_time = time.time()

    if build_destination.exists():
        if rm_destination:
            shutil.rmtree(build_destination)
        else:
            raise DestinationAlreadyExists(build_destination)

    build_destination.mkdir(parents=True)

    if dist_destination.exists():
        if rm_destination:
            shutil.rmtree(dist_destination)
        else:
            raise DestinationAlreadyExists(dist_destination)

    dist_destination.mkdir(parents=True)
    package_name = name
    extra = {}
    if requires_python is not None:
        extra["requires_python"] = requires_python
    package_version = PackageMetadata(
        sdist=not no_sdist, wheel=not no_wheel, wheel_tags=wheel_tags, **extra
    )
    module_name = package_name.replace("-", "_")

    template_config = load_template_config("package")

    logger.debug("Generating project for '%s-%s'", package_name, version)

    package_destination = create_from_template(
        build_destination,
        template_name="package",
        variables={
            "package-name": package_name,
            "module-name": module_name,
            "version": version,
            "dependencies": [],
            "optional-dependencies": [],
            "requires-python": package_version.requires_python,
            "description": package_version.description,
        },
    )

    logger.info(
        "Generated project for '%s-%s'in %.2fms",
        package_name,
        version,
        (time.time() - start_time) * 1000.0,
    )

    for dist in build_package_distributions(
        template_config, package_version, package_destination
    ):
        shared_path = dist_destination / dist.name
        logger.info("Linked distribution to %s", shared_path.relative_to(work_dir))
        shared_path.hardlink_to(dist)

    logger.info(
        "Done in %.2fms",
        (time.time() - start_time) * 1000.0,
    )
