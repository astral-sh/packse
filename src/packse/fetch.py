import logging
import shutil
import subprocess
import tempfile
import time
from pathlib import Path

import pkg_resources

from packse.error import DestinationAlreadyExists

logger = logging.getLogger(__name__)


def fetch(
    dest: Path | None = None,
    ref: str | None = None,
    repo_url: str = "https://github.com/astral-sh/packse",
    repo_subdir: str = "scenarios",
    force: bool = False,
):
    start_time = time.time()
    dest = dest or (Path.cwd() / "scenarios")

    if dest.exists() and not force:
        raise DestinationAlreadyExists(dest)

    if not ref:
        ref = pkg_resources.get_distribution("packse").version
        if ref == "0.0.0":
            ref = "HEAD"

    with tempfile.TemporaryDirectory() as tmpdir:
        # Perform a sparse checkout where we only grab the `scenarios` folder
        logger.info("Cloning repository %s", repo_url)
        subprocess.check_call(
            [
                "git",
                "clone",
                "-n",
                "--filter=tree:0",
                repo_url,
                "repo",
            ],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=tmpdir,
        )

        logger.info("Checking out directory '%s' at ref %s", repo_subdir, ref)
        subprocess.check_call(
            ["git", "sparse-checkout", "set", "--no-cone", repo_subdir],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )
        subprocess.check_call(
            ["git", "fetch", "origin", ref],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )
        subprocess.check_call(
            ["git", "checkout", ref],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )

        src = Path(tmpdir) / "repo" / repo_subdir
        file_count = 0
        for file in sorted(src.iterdir()):
            file_count += 1
            logger.info("Found %s", file.name)

        logger.info("Copying files into '%s'", dest)
        shutil.copytree(src, dest, dirs_exist_ok=True)

    logger.info(
        "Fetched %s files in %.2fs",
        file_count,
        time.time() - start_time,
    )
