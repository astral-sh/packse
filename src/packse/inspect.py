"""
Get details for all scenarios.
"""
import logging
from pathlib import Path
from typing import cast

from packse.error import FileNotFound, InvalidScenario
from packse.scenario import (
    Requirement,
    Scenario,
    load_scenarios,
    scenario_hash,
)
from packse.view import dependency_tree

logger = logging.getLogger(__name__)


def inspect(
    targets: list[Path],
    skip_invalid: bool = False,
    short_names: bool = False,
    no_hash: bool = False,
):
    scenarios_by_path: dict[Path, list[Scenario]] = {}

    # Validate and collect all targets first
    for target in sorted(targets):
        if not target.exists():
            raise FileNotFound(target)

        if target.is_dir():
            for target in target.glob("*.json"):
                try:
                    logger.debug("Loading %s", target)
                    scenarios_by_path[target] = load_scenarios(target)
                except Exception as exc:
                    invalid = InvalidScenario(target, reason=str(exc))
                    if skip_invalid:
                        print(f"Skipping file: {invalid}")
                    else:
                        raise invalid
        else:
            try:
                logger.debug("Loading %s", target)
                scenarios_by_path[target] = load_scenarios(target)
            except Exception as exc:
                invalid = InvalidScenario(target, reason=str(exc))
                if skip_invalid:
                    print(f"Skipping file: {invalid}")
                else:
                    raise invalid

    # Collect a JSON-compatible representation for each scenario
    result = {"scenarios": []}
    for source, scenarios in scenarios_by_path.items():
        for scenario in scenarios:
            scenario = cast(Scenario, scenario)

            raw = scenario.dict()
            raw["source"] = str(source)
            raw["version"] = scenario_hash(scenario)
            raw["tree"] = dependency_tree(scenario).splitlines()
            result["scenarios"].append(raw)

            raw["resolver_options"]["no_build"] = [
                str(
                    Requirement(name).with_unique_name(
                        scenario,
                        raw["version"],
                        short_names,
                        no_hash=no_hash,
                    )
                )
                for name in raw["resolver_options"]["no_build"]
            ]
            raw["resolver_options"]["no_binary"] = [
                str(
                    Requirement(name).with_unique_name(
                        scenario,
                        raw["version"],
                        short_names,
                        no_hash=no_hash,
                    )
                )
                for name in raw["resolver_options"]["no_binary"]
            ]

            # Convert dictionaries to lists for easier templating
            raw["packages"] = [
                {
                    "name": str(
                        Requirement(name).with_unique_name(
                            scenario,
                            raw["version"],
                            short_names,
                            no_hash=no_hash,
                        )
                    ),
                    "versions": [
                        {
                            **package_metadata.dict(),
                            "version": version,
                            "requires": [
                                str(
                                    requirement.with_unique_name(
                                        scenario,
                                        raw["version"],
                                        short_names,
                                        no_hash=no_hash,
                                    )
                                )
                                for requirement in package_metadata.requires
                            ],
                            "extras": [
                                {
                                    "name": extra,
                                    "requires": [
                                        str(
                                            requirement.with_unique_name(
                                                scenario,
                                                raw["version"],
                                                short_names,
                                                no_hash=no_hash,
                                            )
                                        )
                                        for requirement in extra_requires
                                    ],
                                }
                                for extra, extra_requires in package_metadata.extras.items()
                            ],
                        }
                        for version, package_metadata in package.versions.items()
                    ],
                }
                for name, package in scenario.packages.items()
            ]
            raw["expected"]["packages"] = [
                {
                    "name": str(
                        Requirement(name).with_unique_name(
                            scenario,
                            raw["version"],
                            short_names,
                            no_hash=no_hash,
                        )
                    ),
                    "version": version,
                }
                for name, version in scenario.expected.packages.items()
            ]
            raw["root"]["requires"] = [
                {
                    "requirement": str(requirement),
                    "name": requirement.name,
                    "module_name": requirement.name.replace("-", "_"),
                }
                for requirement in (
                    requirement.with_unique_name(
                        scenario,
                        raw["version"],
                        short_names,
                        no_hash=no_hash,
                    )
                    for requirement in scenario.root.requires
                )
            ]

            # Ensure a module name is available for testing import of packages
            for package in raw["expected"]["packages"]:
                package["module_name"] = package["name"].replace("-", "_")

            raw["module_name"] = raw["name"].replace("-", "_")

    return result
