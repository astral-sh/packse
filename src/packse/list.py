"""
List all scenarios.
"""
import logging
from pathlib import Path
from typing import cast

from packse.error import FileNotFound, InvalidScenario
from packse.scenario import (
    Scenario,
    load_scenarios,
    scenario_hash,
)

logger = logging.getLogger(__name__)


def list(
    targets: list[Path],
    no_hash: bool = False,
    skip_invalid: bool = False,
    no_sources: bool = False,
    short_names: bool = False,
):
    scenarios_by_path: dict[Path, list[Scenario]] = {}

    # Validate and collect all targets first
    for target in sorted(targets):
        if not target.exists():
            raise FileNotFound(target)

        try:
            logger.debug("Loading %s", target)
            scenarios_by_path[target] = load_scenarios(target)
        except Exception as exc:
            if not skip_invalid:
                raise InvalidScenario(target, reason=str(exc)) from exc

    # Then list each one
    for source, scenarios in scenarios_by_path.items():
        prefix = "" if no_sources else " " * 4
        if not no_sources:
            print(to_display_path(source))

        for scenario in scenarios:
            scenario = cast(Scenario, scenario)

            if no_hash:
                name = scenario.name
            else:
                name = scenario_hash(scenario)
                if not short_names:
                    name = f"{scenario.name}-{name}"

            print(prefix + name)


def to_display_path(path: Path | str, relative_to: Path | str | None = None) -> str:
    """
    Convert a path to a displayable path. The absolute path or relative path to the
    current (or given) directory will be returned, whichever is shorter.
    """
    path, relative_to = (
        Path(path).resolve(),
        Path(relative_to or ".").resolve(),
    )

    absolute_path = str(path)

    try:
        relative_path = str(path.relative_to(relative_to))
    except ValueError:
        # Not a child path, use the other one
        return absolute_path

    return relative_path if len(relative_path) < len(absolute_path) else absolute_path
