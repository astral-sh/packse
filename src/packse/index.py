import errno
import hashlib
import importlib
import logging
import os
import shutil
import signal
import subprocess
import sys
import tempfile
import time
from contextlib import contextmanager, nullcontext
from pathlib import Path
from typing import Generator

from packse import __development_base_path__
from packse.build import build
from packse.error import (
    RequiresExtra,
    ServeAddressInUse,
    ServeAlreadyRunning,
    ServeCommandError,
)
from packse.inspect import inspect
from packse.template import create_from_template

logger = logging.getLogger(__name__)


def index_up(
    host: str = "localhost",
    port: int = 3141,
    background: bool = False,
    dist_dir: Path | None = None,
    reset: bool = False,
    offline: bool = False,
):
    try:
        importlib.import_module("pypiserver")
    except ImportError:
        raise RequiresExtra("index commands", "index")

    if not dist_dir:
        dist_dir = Path.cwd() / "dist"
    else:
        dist_dir = Path(dist_dir).resolve()

    if background:
        if pid := get_server_pid():
            if is_running(pid):
                raise ServeAlreadyRunning()

    if reset:
        logger.info("Clearing existing server data...")
        if dist_dir.exists():
            shutil.rmtree(dist_dir)

    state_path = get_packse_state_path()

    logger.debug("Storing server metadata at %s", state_path)
    server_log_path = state_path / "server.log" if background else None
    index_url = f"http://{host}:{port}"
    index_note = "offline index" if offline else "local index with PyPI fallback"
    logger.info("Starting %s at %s", index_note, index_url)
    with start_index_server(
        dist_dir,
        host,
        port,
        server_log_path,
        offline=offline,
    ) as server_process:
        # Copy build dependencies over
        add_vendored_build_deps(dist_dir, offline)

        if background:
            logger.info("Running in background with pid %s", server_process.pid)
            logger.info("Stop index server with `packse index down`.")
            print(index_url)

        else:
            debug = logger.getEffectiveLevel() <= logging.DEBUG
            if not debug:
                logger.info("Hiding server output, use `-v` to stream server logs")

            logger.info("Ready! [Stop with Ctrl-C]")

            line = ""
            while True:
                line = server_process.stdout.readline().decode()
                if debug:
                    print(line, end="")


def index_down() -> bool:
    state_path = get_packse_state_path(create=False)

    if not state_path.exists():
        logger.warning("No server detected!")
        return False

    if pid := get_server_pid():
        if not is_running(pid):
            logger.info("Server looks shutdown already.")
            reset_pidfile()
            return

        logger.info("Stopping server with pid %s...", pid)
        try:
            os.kill(pid, signal.SIGINT)
        except ProcessLookupError:
            # Just move on if it's already dead
            pass

        # Wait for the server to exit for 10s
        limit = 10
        start_time = time.time()
        while time.time() - start_time < limit:
            if not is_running(pid):
                break
            time.sleep(0.1)
        else:
            # After the limit is reached, send a stronger signal...
            logger.info("Server has not stopped after %ds, sending kill...", limit)
            os.kill(pid, signal.SIGKILL)

            while is_running(pid):
                time.sleep(0.1)

        reset_pidfile()
        logger.info("Stopped server!")
        return True
    else:
        logger.warning("No server detected!")
        return False


@contextmanager
def start_index_server(
    dist_dir: Path,
    host: str,
    port: int,
    server_log_path: Path | None,
    offline: bool,
) -> Generator[subprocess.Popen, None, None]:
    server_url = f"http://{host}:{port}"
    dist_dir.mkdir(parents=True, exist_ok=True)

    command = [
        sys.executable,
        "-m",
        "pypiserver",
        "run",
        "--host",
        host,
        "--port",
        str(port),
        "-v",
        # Disable auth
        "-P",
        ".",
        "-a",
        ".",
        str(dist_dir),
    ]

    if offline:
        command.append("--disable-fallback")

    logger.debug("Running: %s", " ".join(command))

    server_process = subprocess.Popen(
        command,
        stderr=subprocess.STDOUT,
        stdout=server_log_path.open("wb") if server_log_path else subprocess.PIPE,
    )

    try:
        # Wait for server to be ready
        with server_log_path.open("rb") if server_log_path else nullcontext(
            server_process.stdout
        ) as server_output:
            line = ""
            lines = []

            # Watch for startup
            # TODO: Use a healthceck instead?
            while "Serving on" not in line and "Listening on" not in line:
                try:
                    server_process.wait(0)
                except subprocess.TimeoutExpired:
                    pass
                else:
                    # The process stopped already
                    if (
                        logger.getEffectiveLevel() > logging.DEBUG
                        and server_output.seekable()
                    ):
                        server_output.seek(0)
                    err_output = (
                        "\n".join(lines) + server_output.read().decode().strip()
                    )
                    raise ServeCommandError("Failed to start server!", err_output)

                line = server_output.readline().decode()
                if logger.getEffectiveLevel() <= logging.DEBUG:
                    print(line, end="")
                elif not server_output.seekable():
                    # Track the lines if we'll never be able to get them back
                    lines.append(line)
                if "Address already in use" in line:
                    raise ServeAddressInUse(server_url)

        # Server started!
        write_server_pid(server_process.pid)

        if server_log_path:
            logger.info("Writing server logs to %s", server_log_path)

        try:
            display_path = f"./{dist_dir.relative_to(Path.cwd())}"
        except Exception:
            display_path = dist_dir

        logger.info(f"Serving distributions from {display_path}")

        yield server_process

    except BaseException as exc:
        server_process.terminate()

        stdout, _ = server_process.communicate(timeout=1)
        if logger.getEffectiveLevel() <= logging.DEBUG:
            print(stdout.decode())

        server_process.kill()

        stdout, _ = server_process.communicate(timeout=1)
        if logger.getEffectiveLevel() <= logging.DEBUG:
            print(stdout.decode())

        # Do not reset the pidfile when a server was already running!
        if not isinstance(exc, ServeAddressInUse):
            reset_pidfile()

        raise


def add_vendored_build_deps(dist_dir: Path, offline: bool):
    build_directory = __development_base_path__ / "vendor" / "build"
    if build_directory.exists():
        logger.debug("Copying vendored build dependencies...")
        shutil.copytree(build_directory, dist_dir / "build", dirs_exist_ok=True)
    else:
        logger.warning("No vendored build dependencies found at %s" % build_directory)
        if offline:
            logger.error(
                "Server has no fallback PyPI access and source distribution builds are likely to fail"
            )


def get_packse_state_path(create: bool = True) -> Path:
    path = Path(os.environ.get("PACKSE_STATE_PATH") or (Path.home() / ".packse"))
    if create:
        path.mkdir(exist_ok=True)
    return path


def get_pidfile_path() -> Path:
    return get_packse_state_path() / "server.pid"


def get_server_pid() -> int | None:
    pidfile = get_pidfile_path()
    if not pidfile.exists():
        return None
    else:
        logger.debug("Found existing pid at %s", pidfile)
        return int(pidfile.read_text())


def write_server_pid(pid: int) -> None:
    pidfile = get_pidfile_path()
    pidfile.write_text(str(pid))


def reset_pidfile():
    pidfile = get_pidfile_path()
    pidfile.unlink(missing_ok=True)


def is_running(pid):
    try:
        os.kill(pid, 0)
    except OSError as err:
        # ESRCH: pid not exist
        if err.errno == errno.ESRCH:
            return False
    return True


def sha256_file(path: Path):
    h = hashlib.sha256()

    with open(path, "rb") as file:
        while True:
            # Reading is buffered, so we can read smaller chunks.
            chunk = file.read(h.block_size)
            if not chunk:
                break
            h.update(chunk)

    return h.hexdigest()


def build_index(
    targets: list[Path],
    no_hash: bool,
    short_names: bool,
    dist_dir: Path | None,
):
    out_path = Path("./index")
    if out_path.exists():
        shutil.rmtree(out_path)

    with tempfile.TemporaryDirectory() as tmpdir:
        tmpdir = Path(tmpdir)
        if not dist_dir:
            logger.info("Building distributions...")
            build(
                targets,
                rm_destination=True,
                skip_root=False,
                short_names=short_names,
                no_hash=no_hash,
                dist_dir=tmpdir / "dist",
                build_dir=tmpdir / "build",
            )
            dist_dir = tmpdir / "dist"

        out_path.mkdir()
        (out_path / "files").mkdir()

        variables = inspect(targets, short_names=short_names, no_hash=no_hash)
        for scenario in variables["scenarios"]:
            for package in scenario["packages"]:
                # Create a new distributions section
                package["dists"] = []
                for version in package["versions"]:
                    for file in Path(
                        dist_dir
                        / (
                            scenario["name"]
                            if no_hash
                            else f"{scenario['name']}-{scenario['version']}"
                        )
                    ).iterdir():
                        if package["name"].replace("-", "_") + "-" + version[
                            "version"
                        ] not in str(file):
                            continue

                        # Include all the version information to start
                        dist = version.copy()
                        # Then add a sha256
                        dist["sha256"] = sha256_file(file)
                        dist["file"] = file.name
                        package["dists"].append(dist)
                        logger.info(
                            "Found distribution %s",
                            file.name,
                        )
                        file.rename(out_path / "files" / file.name)

        logger.info("Populating template...")
        create_from_template(
            out_path,
            "index",
            variables=variables,
        )

    logger.info("Built static index at ./%s", out_path)
