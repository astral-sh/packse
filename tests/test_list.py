import pytest
from packse import __development_base_path__

from .common import snapshot_command


def test_list_no_target_finds_all_valid_scenarios(snapshot):
    assert (
        snapshot_command(["list"], snapshot_stdout=False, snapshot_stderr=False)
        == snapshot
    )


def test_list_no_hash(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert snapshot_command(["list", str(target), "--no-hash"]) == snapshot


def test_list_short_names(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert snapshot_command(["list", str(target), "--short-names"]) == snapshot


def test_list_no_sources(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert snapshot_command(["list", str(target), "--no-sources"]) == snapshot


def test_list_target_does_not_exist(snapshot):
    assert snapshot_command(["list", "foo"]) == snapshot


@pytest.mark.usefixtures("tmpcwd")
def test_list_one_target_does_not_exist(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert snapshot_command(["list", str(target), "foo"]) == snapshot


def test_list_invalid_target(snapshot, tmpcwd):
    bad_target = tmpcwd / "test.json"
    bad_target.touch()
    good_target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert snapshot_command(["list", str(bad_target), str(good_target)]) == snapshot


def test_list_invalid_target_skip_invalid(snapshot, tmpcwd):
    bad_target = tmpcwd / "test.json"
    bad_target.touch()
    good_target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert (
        snapshot_command(["list", str(bad_target), str(good_target), "--skip-invalid"])
        == snapshot
    )
