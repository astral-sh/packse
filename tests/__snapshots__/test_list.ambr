# serializer version: 1
# name: test_list_invalid_target
  dict({
    'exit_code': 1,
    'stderr': '''
      File at '[PWD]/test.json' is not a valid scenario: Input data was truncated.
  
    ''',
    'stdout': '',
  })
# ---
# name: test_list_invalid_target_skip_invalid
  dict({
    'exit_code': 0,
    'stderr': '',
    'stdout': '''
      [PROJECT_ROOT]/scenarios/examples/example.json
          example-961b4c22
  
    ''',
  })
# ---
# name: test_list_no_hash
  dict({
    'exit_code': 0,
    'stderr': '',
    'stdout': '''
      scenarios/examples/example.json
          example
  
    ''',
  })
# ---
# name: test_list_no_sources
  dict({
    'exit_code': 0,
    'stderr': '',
    'stdout': '''
      example-961b4c22
  
    ''',
  })
# ---
# name: test_list_no_target_finds_all_valid_scenarios
  dict({
    'exit_code': 0,
    'stderr': '<not included>',
    'stdout': '<not included>',
  })
# ---
# name: test_list_one_target_does_not_exist
  dict({
    'exit_code': 1,
    'stderr': '''
      File 'foo' not found.
  
    ''',
    'stdout': '',
  })
# ---
# name: test_list_short_names
  dict({
    'exit_code': 0,
    'stderr': '',
    'stdout': '''
      scenarios/examples/example.json
          961b4c22
  
    ''',
  })
# ---
# name: test_list_target_does_not_exist
  dict({
    'exit_code': 1,
    'stderr': '''
      File 'foo' not found.
  
    ''',
    'stdout': '',
  })
# ---
